### ISSUE TITLE
[Docs] Install from private git repository with token

### ISSUE THREAD

**Issue by tom-mcclintock** [python-poetry/poetry#7381]
## Feature Request
Relates to #835. Other recent, potentially useful conversations include e.g. #6629.

TLDR - I would like to be able to `poetry install` an environment that includes a package found in a private github repo from within a github action in another private repository (e.g. CI), and use a token instead of a user/pw for credentials. I was not able to find an example of this in the docs.

Example - I have a local project and in my `pyproject.toml` I have a private dependency that looks like this:
```toml
my-private-package = {git = "git@github.com:my-org/my-private-package.git"}
```
Since I have SSH credentials locally, I can `poetry install` and obtain this package no problem.

However, I found the [poetry docs](https://python-poetry.org/docs/dependency-specification/#git-dependencies) lacking in examples for how to do this from a location that does not have SSH credentials. In my case, I am trying to install `my-private-package` from within a github action of some other project.

[This comment from #835](https://github.com/python-poetry/poetry/issues/835#issuecomment-753381859) suggests using a token to obtain permissions at  the `poetry add` step, but does not solve my issue, because `my-private-repo` is already in the `pyproject.toml` file in question (so `poetry add` isn't relevant).

I ended up not using poetry to install the private code (I stuck it in an `optional` group in `pyproject.toml`), and instead used `pip` within the venv:
```yaml
      - name: Activate poetry venv and install private package
        run:  |
          source .venv/bin/activate
          pip install git+https://my-org:${{ secrets.SOME_TOKEN}}@github.com/my-org/my-private-package.git
```
This is not ideal, because it means my local install pattern is different than the pattern used in the action.

I'd like to request this example be formalized and added to the poetry docs, since it seems from the discussions in #835 that it should be possible, but I don't have enough github acumen to figure out how.

Big thanks!

**Comment by madeinoz67** [comment]
have this same issue and came across this.  Thanks have had to use this non-ideal workaround.

I have also attempted ad ding a secondary repository however am seeing the packat appended to the URL and hence can't find it.

A possible solution is using an existing stored keyring credential similar to that of how secondary repository credentials are stored, and maybe a definition of private-package rather than a private-repository to differentiate between the two?

**Comment by elliott-king** [comment]
Another workaround is to have a [git SSH key](https://docs.github.com/en/authentication/connecting-to-github-with-ssh/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent) dedicated to the task. Then you can do something awkward in your container like:

```bash
mkdir /root/.ssh && \
    echo "$GITHUB_SSH_PRIVATE_KEY" > /root/.ssh/id_rsa && \
    chmod 600 /root/.ssh/id_rsa && \
    echo "Host *" > ~/.ssh/config && echo " StrictHostKeyChecking no" >> ~/.ssh/config && \
    poetry install && \
    rm -rf /root/.ssh/
```

If the SSH key is there, poetry will install without an issue.

**Comment by alex-linx** [comment]
We are also facing this issue at our organizations, any solution to allow using poetry install with the github token from github actions?

**Comment by j-gooding** [comment]
From the documentation, can we at least make the following `token` parameter take a file as an argument?
```shell
poetry config repositories.git-org-project https://github.com/org/project.git
poetry config http-basic.git-org-project username token
poetry add git+https://github.com/org/project.git
```
If we can make `token` be a file then we can use docker build time secrets and not have it be leaked in the image. I tried looking through the source code quickly, but it is anything but simple and I don't really have the time to spend hours just learning the code to see if something is even possible or not.

As long as the private git repo is https accessible this should work safely for things like github actions and docker builds.

Is there any more documentation on the relation between `git`, `org`, and `project` in the `config` commands and the `add` command? Do they have to be a one to one mapping? What happens if `org` or `project` contain dashes?

**Comment by wt-asw** [comment]
We have also bumped into this, is there a timeline for this feature to be added?

**Comment by alex-linx** [comment]
A workaround I found for this is to set the url like this:
```
REPO_URL="https://oauth2:${{ inputs.github_token }}@github.com/${GITHUB_REPOSITORY}.git"
```
I know this works for github, not sure about others.

I had a github  action that would patch the pyproject.toml file replacing the url, before building the docker:
```
REPO_URL="https://oauth2:${{ inputs.github_token }}@github.com/${GITHUB_REPOSITORY}.git"
sed -i "s#my-package = .*#my-package = { git = \"$REPO_URL\" }#g" pyproject.toml
```

However this has a drawback that the token is now baked into the image.

In my case I added a command inside the dockerfile that deleted the files that contained it:
```
        rm ./pyproject.toml && \
        rm -f /venv/.venv/lib/python3.9/site-packages/my-package-*.dist-info/direct_url.json; \
```
But this means you can't run poetry install again in the image after the initial install.

**Comment by Scrashdown** [comment]
We have also bumped into this issue in my organization.

In our case, we also need to specify a specific git revision of the package to download. It looks like it is [already possible](https://python-poetry.org/docs/cli/#add) with `poetry add git+<repo url>#<commit sha>`. So to summarize it would be nice if one could pass an auth token **and** a commit SHA to specify which version to download.

**Comment by nitsujri** [comment]
> Another workaround is to have a [git SSH key](https://docs.github.com/en/authentication/connecting-to-github-with-ssh/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent) dedicated to the task. Then you can do something awkward in your CI container like:
> 
> ```shell
> mkdir /root/.ssh && \
> echo "$GITHUB_SSH_PRIVATE_KEY" > /root/.ssh/id_rsa && \
> chmod 600 /root/.ssh/id_rsa && \
> echo "Host *" > ~/.ssh/config && echo " StrictHostKeyChecking no" >> ~/.ssh/config && \
> poetry install && \
> rm -rf /root/.ssh/
> ```
> 
> EDIT: since it's all one line, the ssh key shouldn't end up in your image layers. Feel free to correct me.

A better option today with [Docker is secrets](https://docs.docker.com/build/building/secrets/).

```dockerfile
RUN --mount=type=secret,id=gh_priv_key,target=/root/.ssh/id_rsa \
    /root/.local/bin/poetry install --without dev --no-root
```

```
docker build --secret id=gh_priv_key,src=path/to/private_key .
```

Purpose built to not show up in image layers nor log outputs.

**Comment by codethief** [comment]
To access private git repositories through HTTPS (as opposed to SSH) I just arrived at this solution:

```shell
export GIT_USER="user"
export GIT_PASS="pass"
git config --global credential.helper '!echo "username=${GIT_USER}"; echo "password=${GIT_PASS}";'

poetry install --no-root
```

`pyproject.toml` contains something like

```toml
[tool.poetry.dependencies]
# …
my-private-package = {git = "https://gitlab.com/path/to/my/repo.git" }
```

Background:
- https://git-scm.com/docs/gitcredentials
- https://stackoverflow.com/questions/8536732/can-i-hold-git-credentials-in-environment-variables

**Comment by codethief** [comment]
@elliott-king Maybe I should have been clear about the role of the variable exports that set the credentials in my code snippet: Those were merely supposed to demonstrate how to get poetry to use credentials provided from outside. 

In Docker you would use inject build credentials using `docker build --secret=…` ([docs](https://docs.docker.com/build/building/secrets/)), in Kaniko you would store the credentials in an .env file in the [special `/kaniko` folder](https://github.com/GoogleContainerTools/kaniko/issues/1505) and source that file in your Dockerfile to make the credentials available to Poetry/git.

**Comment by abn** [comment]
Using http auth for git repositories is described in https://python-poetry.org/docs/dependency-specification/#git-dependencies with examples. 

If you want to use git-credentials you will have to use the system git client as described in the doc. But do this only if you really need it. 

If there is a specific question around this please feel free to raise a discussion or ask on our discord server.

**Comment by codethief** [comment]
@abn The suggestion from the page you linked to use `poetry config http-basic.git-org-project username token` is a rather bad solution, though, because it will persist the credentials in the Poetry config and hence in the image layer. In Docker you typically want to use build secrets (see [my previous comment](https://github.com/python-poetry/poetry/issues/7381#issuecomment-1920239247)), i.e. environment variables that don't get persisted.

**Comment by abn** [comment]
@codethief thank you for sharing you POV.

Please see https://python-poetry.org/docs/repositories#configuring-credentials for information on how you can use environment variables or in your case build args.

For a desktop environments (and I guess in container environments were secret storage is enabled) use case Poetry uses keyring to store credentials into your secret store rather than persisting it on file.

Furthermore, you could potentially also rely on multi stage container buidls to only have our virtualenv and application persisted in your runtime container build.

**Comment by codethief** [comment]
> Please see https://python-poetry.org/docs/repositories#configuring-credentials for information on how you can use environment variables or in your case build args.

Thanks, I remember trying `export POETRY_PYPI_TOKEN_FOO=…` a while ago and it turned out to not be an option for me but I don't remember why.

> Furthermore, you could potentially also rely on multi stage container buidls to only have our virtualenv and application persisted in your runtime container build.

In our case, we make heavy use of remote caching, so all image layers (even of intermediate stages) get  in the container registry.
